stock_picking_auto_create_lot/models/stock_move_line.py

```diff
@@ -13,20 +18,50 @@ def _prepare_auto_lot_values(self):
         Prepare multi valued lots per line to use multi creation.
         """
         self.ensure_one()
-        return {"product_id": self.product_id.id, "company_id": self.company_id.id}
+        vals = {"product_id": self.product_id.id, "company_id": self.company_id.id}
+        if self.product_id.tracking == "lot":
+            expression = self.picking_type_id.auto_create_lot_name_expression
+            if expression:
+                template_instructions = parse_inline_template(expression)
+                variables = {"stock_move_line": self}
+                variables.update(template_env_globals)
+                vals["name"] = render_inline_template(template_instructions, variables)
+        return vals
 
-    def set_lot_auto(self):
+    def _get_and_create_lots(self):
         """
-        Create lots using create_multi to avoid too much queries
-        As move lines were created by product or by tracked 'serial'
-        products, we apply the lot with both different approaches.
+        Get existing lots and create missing ones.
         """
-        values = []
+        vals_list = []
         stock_lot_obj = self.env["stock.lot"]
-        lots_by_product = dict()
+        lots = stock_lot_obj.browse()
         for line in self:
-            values.append(line._prepare_auto_lot_values())
-        lots = stock_lot_obj.create(values)
+            vals = line._prepare_auto_lot_values()
+            if line.picking_id.use_existing_lots and vals.get("name"):
+                lot = stock_lot_obj.search(
+                    [
+                        (key, "=", vals[key])
+                        for key in ("product_id", "company_id", "name")
+                    ]
+                )
+                if lot:
+                    lots |= lot
```

Comment:
```suggestion
                lot = stock_lot_obj.search(
                    [
                        (key, "=", vals[key])
                        for key in ("product_id", "company_id", "name")
                    ],
                    limit=1
                )
                if lot:
                    lot_ids |= lot.id
```

Reply:
Alwasy use limit=1 if you expect to have only 1 result and avoid interpolating recordsets if you think you'll have many records there (collect ids then browse at the end).

Reply:
@simahawk thanks a lot for your review. i get and agree with the `limit=1`, but i don’t understand why it is better to collect ids then browse for the records instead of getting the records directly and adding them to a record set. where does the performance gain come from?

Reply:
thanks, i learned a lot thanks to you! until now, i assumed that `.browse([ids])` performed a database query to populate the recordset, but it just creates the recordset without doing any query.

is what you call “interpolating” using the `|=` operator? i checked, and you’re right, doing this is heavier. it calls [`.union()`](https://github.com/OCA/OCB/blob/16.0/odoo/models.py#L5830) which checks for the record types and calls `.browse()`, so calling `.browse()` once at the end is more optimal. however, this seems like a very small performance gain when compared to doing a database query, for example. did you encounter cases where the gain is significant?

i added a new commit with your suggested changes. is it what you meant? i thinks it makes the code slightly less readable, no?

Reply:
less readable: I don't think so :)
Anyway, thank for your followups  and sorry for being later here :)